subsys/net/conn_mgr/online_checker.c

```diff
@@ -0,0 +1,559 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(online_checker, CONFIG_NET_CONNECTION_MANAGER_LOG_LEVEL);
+
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/random/random.h>
+#include <errno.h>
+#include <zephyr/net/net_core.h>
+#include <zephyr/net/net_if.h>
+#include <zephyr/net/icmp.h>
+#include <zephyr/net/socket.h>
+#include <zephyr/net/net_mgmt.h>
+#include <zephyr/net/conn_mgr_monitor.h>
+#include <zephyr/net/http/client.h>
+#include "conn_mgr_private.h"
+#include "net_private.h"
+
+#if defined(CONFIG_NET_CONNECTION_MANAGER_ONLINE_CHECK_PING)
+#define PING_HOST CONFIG_NET_CONNECTION_MANAGER_ONLINE_CHECK_PING_HOST
+#else
+#define PING_HOST ""
+#endif
+
+#if defined(CONFIG_NET_CONNECTION_MANAGER_ONLINE_CHECK_HTTP)
+#define ONLINE_CHECK_URL CONFIG_NET_CONNECTION_MANAGER_ONLINE_CHECK_HTTP_URL
+#define MAX_RECV_BUF_LEN 512
+#else
+#define ONLINE_CHECK_URL ""
+#define MAX_RECV_BUF_LEN 1
+#endif
+
+#define WAIT_TIMEOUT CONFIG_NET_CONNECTION_MANAGER_ONLINE_CHECK_TIMEOUT
+#define MAX_HOSTNAME_LEN CONFIG_NET_CONNECTION_MANAGER_ONLINE_CHECK_MAX_HOSTNAME_LEN
+
+struct online_check_data {
+	net_conn_mgr_online_checker_t cb;
+	void *user_data;
+	char *host; /* This points to hostname_port */
+	char *port; /* and same for this */
+	struct sockaddr hostaddr;
+	char hostname_port[MAX_HOSTNAME_LEN];
+	uint8_t http_recv_buf[MAX_RECV_BUF_LEN];
+	bool hostaddr_valid : 1;
+	bool is_tls : 1;
+};
+
+static struct online_check_data online_check_data_storage;
+static struct online_check_data *online_check = &online_check_data_storage;
+
+static enum dns_resolve_status resolve_host(const char *host,
+					    const char *service,
+					    int socktype,
+					    struct zsock_addrinfo **res)
+{
+	struct zsock_addrinfo hints;
+	int ret;
+
+	if (!host || host[0] == '\0') {
+		NET_WARN("Online check hostname missing.");
+		return DNS_EAI_NONAME;
+	}
+
+	memset(&hints, 0, sizeof(hints));
+	if (IS_ENABLED(CONFIG_NET_IPV4) && !IS_ENABLED(CONFIG_NET_IPV6)) {
+		hints.ai_family = AF_INET;
+	} else if (!IS_ENABLED(CONFIG_NET_IPV4) && IS_ENABLED(CONFIG_NET_IPV6)) {
+		hints.ai_family = AF_INET6;
+	} else {
+		hints.ai_family = AF_UNSPEC;    /* Allow IPv4 or IPv6 */
+	}
+
+	hints.ai_socktype = socktype;
+
+	NET_DBG("Resolving %s", host);
+
+	ret = zsock_getaddrinfo(host, service, &hints, res);
+	if (ret != 0) {
+		NET_DBG("Cannot get %s (%d/%s)", host,
+			ret, zsock_gai_strerror(ret));
+		return ret;
+	}
+
+	return 0;
+}
+
+static int icmp_handler(struct net_icmp_ctx *ctx,
```

Comment:
Can we simplify the overall architecture, and have there be a user-defined callback that just spits back an IP address and port?

Custom IP resolution strategies are unfortunately commonly necessary anyways